x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer;
+
+import org.elasticsearch.xpack.esql.expression.function.scalar.math.AutoBucket;
+import org.elasticsearch.xpack.esql.optimizer.OptimizerRules.LogicalPlanDependencyCheck;
+import org.elasticsearch.xpack.ql.common.Failure;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.plan.logical.LogicalPlan;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.ql.common.Failure.fail;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FOURTH;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.THIRD;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isFoldable;
+
+public final class LogicalVerifier {
+
+    private static final LogicalPlanDependencyCheck DEPENDENCY_CHECK = new LogicalPlanDependencyCheck();
+    public static final LogicalVerifier INSTANCE = new LogicalVerifier();
+
+    private LogicalVerifier() {}
+
+    /** Verifies the optimized logical plan. */
+    public Collection<Failure> verify(LogicalPlan plan) {
+        Set<Failure> failures = new LinkedHashSet<>();
+
+        plan.forEachUp(p -> {
+            // dependency check
+            DEPENDENCY_CHECK.checkPlan(p, failures);
```

Comment:
Hmm, this means we treat optimization errors like other verification failures - they become client (400) exceptions (`VerificationException`). I don't think this is correct.

Shouldn't we rather throw an `IllegalStateException` if the dependency check fails?

Also applies to the PhysicalVerifier.

Reply:
let's take this offline.

Reply:
Indeed, I think we should treat these planning inconsistencies different than throwing an `EsqlClientException`.